### PR TITLE
Fix the bug that PollOwnerForActorOutOfScope's reply may be missing

### PR DESCRIPTION
**Pull request by ffbin**
## Why are these changes needed?
Problem:
GCS server may send a WaitForActorOutOfScope request to CoreWorker after reference counter releases the actor object and GCS server will not be able to receive a response to the WaitForActorOutOfScope request.
Solution:
When a WaitForActorOutOfScope request is received, we will respond immediately if the actor object has already been evicted, so that the gcs server can destroy the actor.

## Related issue number
https://github.com/ray-project/ray/issues/9267

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/core_worker/core_worker.cc [hunk 1]
```diff
@@ -1812,7 +1799,20 @@ void CoreWorker::HandleWaitForActorOutOfScope(
   if (it == actor_handles_.end()) {
     respond(actor_id);
   } else {
-    RAY_CHECK(actor_out_of_scope_callbacks_.emplace(actor_id, std::move(respond)).second);
```
Comment (rkooo567) [inline src/ray/core_worker/core_worker.cc:1815]: Don't we still need this when gcs actor management is off?

File: python/ray/tests/test_reference_counting.py [hunk 2]
```diff
@@ -527,6 +527,42 @@ def test_basic_nested_ids(one_worker_100MiB):
...
+    @ray.remote
+    class A:
+        pass
+
+    a = A.remote()
+    b = A.remote()
+
+    del a
+    del b
+    time.sleep(10)
```
Comment (rkooo567) [inline python/ray/tests/test_reference_counting.py:558]: Can we avoid using time.sleep? We can probably just wait until `ray.actors()` return the desired values.

Comment (stephanie-wang) [reply python/ray/tests/test_reference_counting.py:558]: Yes, please use a wait with an error on timeout instead.

File: src/ray/core_worker/core_worker.cc [hunk 3]
```diff
@@ -1812,7 +1799,20 @@ void CoreWorker::HandleWaitForActorOutOfScope(
...
+      };
+
+      // Returns true if the object was present and the callback was added. It might have
+      // already been evicted by the time we get this request, in which case we should
+      // respond immediately so the gcs server can destroy the actor.
+      const auto actor_creation_return_id = ObjectID::ForActorHandle(actor_id);
+      if (!reference_counter_->SetDeleteCallback(actor_creation_return_id, callback)) {
+        RAY_LOG(DEBUG) << "ActorID reference already gone for " << actor_id;
+        respond(actor_id);
+      }
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc:1814]: Nice, this looks much better.

File: src/ray/core_worker/core_worker.cc [hunk 4]
```diff
@@ -1812,7 +1799,20 @@ void CoreWorker::HandleWaitForActorOutOfScope(
   if (it == actor_handles_.end()) {
     respond(actor_id);
   } else {
-    RAY_CHECK(actor_out_of_scope_callbacks_.emplace(actor_id, std::move(respond)).second);
+    if (RayConfig::instance().gcs_actor_service_enabled()) {
```
Comment (stephanie-wang) [inline src/ray/core_worker/core_worker.cc:1802]: Can you add a comment to say why we don't need to do this if the GCS actor service is disabled?

**Review by stephanie-wang — APPROVED** [review]
Looks great!

File: src/ray/core_worker/core_worker.cc (hunk 1 above)
Comment (ffbin) [reply src/ray/core_worker/core_worker.cc]: actor_out_of_scope_callbacks_ is only used when gcs actor management is turn on.

File: python/ray/tests/test_reference_counting.py (hunk 2 above)
Comment (ffbin) [reply python/ray/tests/test_reference_counting.py]: fixed

File: src/ray/core_worker/core_worker.cc (hunk 4 above)
Comment (ffbin) [reply src/ray/core_worker/core_worker.cc]: fixed

File: python/ray/tests/test_reference_counting.py [hunk 5]
```diff
@@ -527,6 +527,51 @@ def test_basic_nested_ids(one_worker_100MiB):
     _fill_object_store_and_get(inner_oid_bytes, succeed=False)
 
 
+def _check_actor_states(expected):
+    actors = list(ray.actors().values())
+    for actor in actors:
+        assert actor["State"] == expected
+
+
+def check_actor_states(expected, timeout=10):
```
Comment (raulchen) [inline python/ray/tests/test_reference_counting.py:536]: You can just use the `wait_for_condition` helper function.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 6]
```diff
@@ -733,6 +718,12 @@ void GcsActorManager::OnActorCreationSuccess(const std::shared_ptr<GcsActor> &ac
         RAY_CHECK(!worker_id.IsNil());
         RAY_CHECK(!node_id.IsNil());
         RAY_CHECK(created_actors_[node_id].emplace(worker_id, actor_id).second);
+
+        if (!actor->IsDetached() && worker_client_factory_) {
+          // This actor is owned. Send a long polling request to the actor's
+          // owner to determine when the actor should be removed.
+          PollOwnerForActorOutOfScope(actor);
```
Comment (raulchen) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:725]: In `PollOwnerForActorOutOfScope`, we should handle the case where the owner has died.

File: python/ray/tests/test_reference_counting.py (hunk 5 above)
Comment (raulchen) [reply python/ray/tests/test_reference_counting.py:536]: something like
```
def _all_actors_dead():
   return all(actor["State"] == DEAD for actor in ray.actors.values())

wait_for_condition(_all_actors_dead, timeout=..)

Comment (ffbin) [reply python/ray/tests/test_reference_counting.py]: fixed

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc (hunk 6 above)
Comment (ffbin) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:733]: fixed

**Review by raulchen — APPROVED** [review]

**Pull request merged**